### PR TITLE
Strip debug symbols + apply upx to reduce bin size

### PR DESCRIPTION
**Pull request by pdalpra**
Seen on https://blog.filippo.io/shrink-your-go-binaries-with-this-one-weird-trick/

Before :
```
$ ls -lh _build 
total 34M
-rwxr-xr-x 1 pdalpra pdalpra 12M Apr 17 22:01 teads-central_darwin_amd64
-rwxr-xr-x 1 pdalpra pdalpra 12M Apr 17 22:01 teads-central_linux_amd64
-rwxr-xr-x 1 pdalpra pdalpra 12M Apr 17 22:01 teads-central_windows_amd64.exe
```

After:
```
$ ls -lh _build   
total 24M
-rwxr-xr-x 1 pdalpra pdalpra 8.3M Apr 17 22:02 teads-central_darwin_amd64
-rwxr-xr-x 1 pdalpra pdalpra 7.7M Apr 17 22:02 teads-central_linux_amd64
-rwxr-xr-x 1 pdalpra pdalpra 7.8M Apr 17 22:02 teads-central_windows_amd64.exe
```

### CODE REVIEW

**Comment by pdalpra** [comment]
Not applying the `upx` trick for the moment, but it's extremely efficient ! 

```
-rwxr-xr-x 1 pdalpra pdalpra 1.6M Apr 17 22:02 teads-central_linux_amd64
```

**Comment by pdalpra** [comment]
Well, it's too damn efficient to just not use it :
```
$ ls -lh _build                           
total 5.1M
-rwxr-xr-x 1 pdalpra pdalpra 1.8M Apr 17 22:02 teads-central_darwin_amd64
-rwxr-xr-x 1 pdalpra pdalpra 1.6M Apr 17 22:02 teads-central_linux_amd64
-rwxr-xr-x 1 pdalpra pdalpra 1.7M Apr 17 22:02 teads-central_windows_amd64.exe
```

As it's quite a slow process, it will be only applied when releasing (where it actually make sense anyway).

**Comment by bjaglin** [comment]
LGTM for stripping, but regarding up , for a CLI meant to be invoked a lot, I am not sure trading off boot time for binary size is really a good thing, is it?

**Comment by pdalpra** [comment]
There is indeed an overhead, there's no denying it.

With UPX:

```
$ time _build/teads-central_linux_amd64 
...
_build/teads-central_linux_amd64  0.16s user 0.00s system 98% cpu 0.169 total
```

Without UPX:
```
$ time _build/teads-central_linux_amd64 
...
_build/teads-central_linux_amd64  0.00s user 0.01s system 84% cpu 0.012 total
```

However, we're talking about a 150ms overhead, for a tool which involves communicating with Docker a lot, waiting for containers, looking for tags on the registry for tens of Docker images, pushing images to a Docker registry, etc...

Additionally, teads-central will reside in the workspace of every job where teads-central will be integrated, which could easily means ~8 MB * 100 or 200 jobs, taking up a lot of space in the end.

In the end, IMHO, the benefit of having a binary `upx`ed which takes about 20% of the original binary size for a 150ms overhead trumps the reduced boot time of the untainted binary for our use cases.

WDYT ?

**Comment by bjaglin** [comment]
There is indeed a tradeoff to find... What would be the overhead in https://github.com/ebuzzing/tool-stack-launcher/pull/7 for example/

**Comment by bjaglin** [comment]
cc @ebuzzing/platform since it's an interesting discussion

**Comment by pdalpra** [comment]
The overhead is fixed AFAIK, so it'd still be 150ms, at least on Linux.
I'll test on Mac OS X too, but I guess the decompression overhead would be approximately the same no matter the OS.

**Comment by bjaglin** [comment]
@pdalpra I understand the overhead is fixed, but I wanted to get a feeling of the actualy overhead as the binary is called many times

**Comment by pdalpra** [comment]
Ah, sorry I misunderstood :)

A small "benchmark":

### Currently

Cold start: 
```
FEATURE_BRANCH=digital-element ./run.sh  0.25s user 0.08s system 3% cpu 9.483 total
```
After a few runs:
```
FEATURE_BRANCH=digital-element ./run.sh  0.19s user 0.06s system 4% cpu 5.993 total
```

### teads-central without downloading, without upx

Cold start: 
```
FEATURE_BRANCH=digital-element ./run.sh  0.67s user 0.09s system 16% cpu 4.730 total
```
After a few runs:
```
FEATURE_BRANCH=digital-element ./run.sh  0.70s user 0.07s system 27% cpu 2.774 total
```

### teads-central, without downloading, with upx

Cold start: 
```
FEATURE_BRANCH=digital-element ./run.sh  0.70s user 0.07s system 27% cpu 6.156 total
```
After a few runs:
```
FEATURE_BRANCH=digital-element ./run.sh  1.12s user 0.10s system 39% cpu 3.096 total
```

**Comment by bjaglin** [comment]
fair enough, the tradeoff is acceptable, LGTM!

**Comment by pdalpra** [comment]
TL;DR for other readers: it seems that once teads-central gets loaded somewhere in RAM, the negative impact of `upx` goes away.
